### ISSUE TITLE
[Additional Feature] Machine Learning Predictive Model

### ISSUE THREAD

**Issue by eo2000** [XavierGuertin/WildfireVisualizationProject#310]
**Description:** 

This issue covers developing a machine learning model as an additional non-critical feature.

Our current platform integrates and visualizes historical/synthetic wildfire data, but it would be interesting to implement and add a feature that utilizes algorithms to forecast fire behaviour based on weather, terrain, and vegetation, to anticipate fire spread.

The goal of this task would be to develop a machine learning predictive model that forecasts wildfire behaviour based on past environmental and meteorological data. This model should integrate with a frontend visualization and provide clear documentation.

Size/Points: Large
Priority: High 
Risk: Medium (data sparsity, accuracy of predictions)
Ideal Time: 1-2 weeks

**User Story:**
As a user,
I want to predict future wildfire behaviours based on past environmental data, so that I can visualize potential wildfire risks and areas of concern in the future. 

**Technical Design:** 
- Dataset – Identify and preprocess relevant datasets.
- Model Development – Define ML architecture, train, and evaluate the model.
- Frontend – Implement visualization for predicted wildfire behaviours. 
- Testing & Validation – Validate predictions and optimize model performance.

**Acceptance Criteria:**
- Model & Data –  Identify a suitable dataset and model for predicting wildfire behaviours with measurable accuracy.
- Frontend Integration - Should integrate with a frontend visualization 
- Performance – Runs efficiently
- Clear documentation

**Comment by eo2000** [comment]
### Dataset Research 
Potential datasets : 
- https://hirondelle.crim.ca/stac
- https://stacindex.org/catalogs/forest-observatory#/
- https://stacindex.org/catalogs/maxar-open-data-catalog-ard-format#/
- https://archive.ics.uci.edu/dataset/162/forest+fires
- https://firms.modaps.eosdis.nasa.gov/download/
- https://open.canada.ca/data/en/dataset/9d8f219c-4df0-4481-926f-8a2a532ca003 / https://www.kaggle.com/datasets/abdelghaniaaba/wildfire-prediction-dataset/data
- https://open.canada.ca/data/en/dataset/1728e7f9-472a-474e-9e04-f96bf59479f9
- https://www.kaggle.com/datasets/fantineh/next-day-wildfire-spread?resource=download

**Comment by eo2000** [comment]
# Engineering design & implementation: 
## **1. Understanding the Requirements**
### **Initial Task or Problem**
Our current platform integrates and visualizes historical/synthetic wildfire data, but it would be interesting to implement and add a feature that utilizes algorithms to forecast fire behaviour based on weather, terrain, and vegetation, to anticipate fire spread.

The goal of this task would be to develop a machine learning predictive model that forecasts wildfire behavior based on past environmental and meteorological data. This model should integrate with a frontend visualization and provide clear documentation.

Following research on existing datasets, and predictive modeling, the following wildfire behaviors can be predicted: wildfire spread, wildfire burnt area, next-likely wildfire.

Initially, the Next Day Wildfire Spread (Kaggle) dataset was considered, but its format required extensive preprocessing. Instead, the Forest Fires Dataset (UCI) was chosen for its structured meteorological data, though it is geographically limited to Portugal. If more Canadian data becomes available, the model can be retrained.


## **2. Exploring Possible Solutions**
### Dataset Research 
Reference this [comment](https://github.com/XavierGuertin/WildfireVisualizationProject/issues/310#issuecomment-2744481728) 


### Solution Design: Machine Learning Model
#### Model Architecture: Support Vector Regression (SVR)
Support Vector Regression (SVR) is a robust machine learning technique that effectively captures non-linear relationships. The SVR model utilizes a Radial Basis Function (RBF) kernel to model complex interactions in the data.

#### Initial Challenges & Model Selection:
Initially, model training performed poorly because as several features had to be removed due to data limitations. Various architectures and models were tested, but they yielded unsatisfactory results. Finally, the SVR model was selected, which significantly improved prediction accuracy.

#### Steps for Model Implementation:
1. Data Preprocessing
    * Loading and cleaning the dataset: Removing missing values from key features (X, Y, month, day, temperature, RH, wind, rain).
    * One-hot encoding categorical variables (month, day) to ensure compatibility with the model.
    * Defining inputs (X) and target (y), with y representing the burned area.
    * Applying log transformation (log(y + 1)) to reduce skewness in the target variable.
2. Model Training and Optimization
    * SVR Motivation & Key Parameters
        * C: Regularization parameter balancing fit and generalization.
        * ε (epsilon): Defines a margin of tolerance where no penalty is applied.
        * Gamma: Controls the impact of individual training samples on predictions.
    * Hyperparameter Tuning
        * GridSearchCV was used to fine-tune gamma values during cross-validation.
    * Training Scheme
        * Outer Loop: 30 iterations of training and testing with shuffled datasets.
        * Inner Loop: 10-fold cross-validation to ensure robustness.
        * Feature Scaling: StandardScaler applied to normalize input features
3. Model Evaluation Metrics
To assess the performance of the wildfire spread prediction model, the following evaluation metrics were used:
  * Mean Absolute Error (MAE): Measures the absolute difference between predicted and actual values.
  * Root Mean Squared Error (RMSE): Penalizes large errors more than MAE, useful for detecting outliers.
  * R² (Coefficient of Determination): Indicates how well the independent variables explain the variation in wildfire spread.
  * Log-Transformed Target Evaluation: Since the model predicts log-transformed values, outputs are exponentiated to return them to their original scale.

Model Performance Summary

| Metric  | Value |
|---------|-----------------|
| MAE     | 12.647 hectares |
| RMSE    | 64.684 hectares |
| R²      | -0.035          |

These results were compared with values reported in the [reference dataset paper](https://www.semanticscholar.org/paper/A-data-mining-approach-to-predict-forest-fires-data-Cortez-Morais/0f529dc2b2b2bad22394454d4cba79e2c319f0b0): 

* Reported MAE: 13.13 ± 0.02 hectares
* Reported RMSE: 64.7 ± 0.0 hectares

These results mean that the model's MAE is slightly better than the reference, the RMSE is basically the same, so large errors are on par with what's expected, and the Negative R² suggests that the model isn't capturing much meaningful variance — despite decent error metrics. See section "Lessons Learned & Recommendations" for recommendations on improving metrics.

The final model, along with the scaler, was saved as 'final_pipeline.joblib' for deployment.

---

## **4. Technical Challenges & Backend Integration**
### **Challenges Faced:**
* Scope of Predictive Model: The predictive model needed to ensure alignment with stakeholder requirements and relevant datasets.
* Quantity of Data: Managing imbalances in training data.
* API Integration: Integrating the trained model via APIs for frontend interaction. 


### **5.Integration with Services:**
- Deployed API endpoints to serve predictions to the frontend.

---

## **6. Frontend Communication & UX Enhancements**
### **Frontend Adaptations:**
Visualized model predictions on platform.

Mockup:

<img width="636" alt="Image" src="https://github.com/user-attachments/assets/ff07374e-c970-4d53-a6ee-543ef159d2d6" />

Frontend Implementation:

<img width="707" alt="Image" src="https://github.com/user-attachments/assets/8c1ce370-1b00-461f-8f5e-77bc2da96e71" />

Predicted Area Visualization on Map:
1. User submits input → form sends values like X, Y, temperature, etc.
2. Prediction is returned from the API as a value in hectares.
3. Coordinates (X, Y) from the grid are converted to real lat/lon using a helper function (gridToLonLat()).
4. The predicted area is converted from hectares to square meters, and then to radius in degrees.
5. A circular geometry is created using ol/geom/Circle, centered at the mapped lat/lon, with the calculated radius.
6. This circle is styled with a semi-transparent red color and added to the map as a vector layer.
7. The map animates and zooms to that location so the burn area is clearly visible.

<img width="1405" alt="Image" src="https://github.com/user-attachments/assets/4bc59f76-0a44-4957-b94a-2f55382575a5" />

---

## **7. Final Thoughts / Engineering Outcome**
### **Results & Benefits:**
- Delivered a scalable and efficient predictive model.
- Improved decision-making support through enhanced visualizations.
- Provided documentation to support future iterations (Wiki Link).

### **Lessons Learned & Recommendations:**
- Incorporating explainable AI techniques increased model trust.
- Future implementations should include data from within Canada and a larger dataset.
- Recommendations for Improving Model Metrics:
1. Feature Engineering Enhancements:
Include derived features such as drought indices, vegetation health (e.g., NDVI), or fire history.
Add interaction terms (e.g., temperature * wind) to capture compound effects.
2. Model Alternatives:
Random Forests or Gradient Boosting Machines (e.g., XGBoost, LightGBM)
Test neural networks, particularly if more data becomes available.
3. Target Engineering:
Consider predicting log(area + 1) directly and optimizing metrics in log space.
4. Outlier Handling:
Many wildfire datasets are highly skewed — using quantile loss, robust scaling, or truncated training ranges can stabilize training and improve R².

**Comment by eo2000** [comment]
[Wiki Page](https://github.com/XavierGuertin/WildfireVisualizationProject/wiki/Predictive-Modelling-%E2%80%90-Burnt-Area-Prediction-Model-Guide) for this new feature

**Comment by eo2000** [comment]
Demo:

https://github.com/user-attachments/assets/4a3f4317-498d-4a98-8c61-fc5aaae04ad3

**Comment by yasminehilout** [comment]
We are unassigning ourselves from this issue as we are officially handing over the capstone project to the stakeholder. This issue remains open for future consideration and development by the stakeholder or future team members.

**Comment by yasminehilout** [comment]
This issue was identified during Release 3 but is not within the final deliverables of our capstone scope.
This issue remains open for future consideration and development by the stakeholder or any future team.
